# Machine Learning Spring 2025

> **Project 1 - Temperature Prediction**<br><br>Team: ST_ML2025_2

In [76]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
import re
import math

In [77]:
plt.rcParams['font.family'] ='NanumGothic'
plt.rcParams['axes.unicode_minus'] =False

In [78]:
os.environ["KMP_DUPLICATE_LIB_OK"] = "True"

In [79]:
train_df = pd.read_csv("input/train_dataset.csv")
test_df = pd.read_csv("input/test_dataset.csv")

# Dataset 분석

주어진 데이터셋의 구성은 다음과 같다.

- `train_dataset.csv`: 동두천, 서울, 강화, 인천, 이천, 양평 관측소의 2019-2024년 데이터셋
- `test_dataset.csv`: 파주, 수원 관측소 데이터셋
- `station_info.csv`: 기상 관측소별 정보

In [80]:
train_df.sample(10)

id  station station_name   date  cloud_cover_0  cloud_cover_1  \
10635  15014      202           양평  02-27            0.0            0.0   
12147  16526      203           이천  04-21            0.0            0.0   
2494    4682      108           서울  11-04            0.0            0.0   
11241  15620      203           이천  10-28            0.0            0.0   
8021   12400      201           강화  12-26            0.0            0.0   
4140    6328      108           서울  05-07           10.0           10.0   
11253  15632      203           이천  11-09            0.0            0.0   
9572   13951      202           양평  03-29            9.0            9.0   
9608   13987      202           양평  05-04            6.0           10.0   
1458    1458       98          동두천  12-30            0.0            0.0   

       cloud_cover_10  cloud_cover_11  cloud_cover_12  cloud_cover_13  ...  \
10635             0.0             0.0             0.0             0.0  ...   
12147             1.0             1.0             0.0             0.0  ...   
2494              0.0             0.0             0.0             0.0  ...   
11241             0.0             0.0             0.0             1.0  ...   
8021              3.0             7.0             7.0             8.0  ...   
4140              9.0             9.0             9.0             7.0  ...   
11253             0.0             0.0             0.0             0.0  ...   
9572              0.0             0.0             0.0             0.0  ...   
9608             10.0            10.0             9.0            10.0  ...   
1458              0.0             0.0             0.0             0.0  ...   

       wind_speed_23  wind_speed_3  wind_speed_4  wind_speed_5  wind_speed_6  \
10635            0.9           0.9           0.7           1.2           1.4   
12147            1.0           0.0           0.3           0.5           0.0   
2494             1.2           2.4           2.5           2.2           1.9   
11241            0.9           0.1           0.4           0.2           0.0   
8021             0.6           0.6           0.7           1.0           0.6   
4140             2.7           1.5           2.1           1.1           1.8   
11253            0.2           0.0           0.3           0.2           0.2   
9572             0.5           3.6           2.6           2.1           1.3   
9608             0.1           0.4           1.0           0.6           0.9   
1458             0.0           0.3           0.6           0.0           0.1   

       wind_speed_7  wind_speed_8  wind_speed_9  climatology_temp    target  
10635           1.5           2.4           1.4          4.578571  0.521429  
12147           0.2           0.4           0.9         15.205357  0.194643  
2494            2.5           1.4           2.2         10.255357  2.144643  
11241           0.1           0.9           0.7         11.508929  0.191071  
8021            0.9           0.4           0.6         -3.485714 -1.714286  
4140            1.9           2.8           3.7         15.046429 -0.346429  
11253           0.0           0.2           0.6          8.169643 -2.469643  
9572            2.2           1.4           1.0          9.332143  1.267857  
9608            0.3           1.3           2.0         16.385714 -1.785714  
1458            0.3           0.3           0.1         -4.487500  0.787500  

[10 rows x 342 columns]

In [81]:
train_df.describe()

id       station  cloud_cover_0  cloud_cover_1  \
count  13132.000000  13132.000000   12945.000000   12920.000000   
mean    9484.110493    153.980658       2.915798       3.022291   
std     5311.954253     48.183220       3.646779       3.652165   
min        0.000000     98.000000       0.000000       0.000000   
25%     5470.750000    108.000000       0.000000       0.000000   
50%     8753.500000    112.000000       0.000000       1.000000   
75%    14227.250000    202.000000       6.000000       6.000000   
max    17510.000000    203.000000      10.000000      10.000000   

       cloud_cover_10  cloud_cover_11  cloud_cover_12  cloud_cover_13  \
count    12916.000000    12926.000000    12931.000000    12926.000000   
mean         3.126742        3.092063        3.046400        3.035046   
std          3.659422        3.589739        3.512063        3.483002   
min          0.000000        0.000000        0.000000        0.000000   
25%          0.000000        0.000000        0.000000        0.000000   
50%          1.000000        1.000000        1.000000        1.000000   
75%          7.000000        7.000000        6.000000        6.000000   
max         10.000000       10.000000       10.000000       10.000000   

       cloud_cover_14  cloud_cover_15  ...  wind_speed_23  wind_speed_3  \
count    12919.000000    12938.000000  ...   13122.000000  13120.000000   
mean         2.992182        2.911115  ...      -0.032754     -2.491814   
std          3.462070        3.440298  ...     123.464434    195.197628   
min          0.000000        0.000000  ...   -9999.000000  -9999.000000   
25%          0.000000        0.000000  ...       0.600000      0.400000   
50%          1.000000        1.000000  ...       1.200000      1.000000   
75%          6.000000        6.000000  ...       2.100000      1.900000   
max         10.000000       10.000000  ...      10.700000     11.300000   

       wind_speed_4  wind_speed_5  wind_speed_6  wind_speed_7  wind_speed_8  \
count  13116.000000  13112.000000  13119.000000  13120.000000  13125.000000   
mean      -0.234637     -1.011882     -4.068389     -7.119787    -10.073272   
std      123.489868    151.258218    230.949420    289.453706    337.895538   
min    -9999.000000  -9999.000000  -9999.000000  -9999.000000  -9999.000000   
25%        0.400000      0.400000      0.400000      0.400000      0.500000   
50%        0.900000      0.900000      0.900000      0.900000      1.000000   
75%        1.900000      1.800000      1.800000      1.800000      2.000000   
max        9.500000     10.600000     12.400000     11.100000     11.300000   

       wind_speed_9  climatology_temp        target  
count  13127.000000      13132.000000  13132.000000  
mean      -9.119029         12.658557      0.221979  
std      326.432094         10.023504      2.960544  
min    -9999.000000         -4.487500    -12.864286  
25%        0.700000          3.292857     -1.643052  
50%        1.300000         12.842857      0.157143  
75%        2.200000         22.271429      2.045536  
max        9.400000         28.455357     11.778571  

[8 rows x 340 columns]

In [82]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13132 entries, 0 to 13131
Columns: 342 entries, id to target
dtypes: float64(338), int64(2), object(2)
memory usage: 34.3+ MB


## Feature 분석

### Feature 분류 및 설명

In [83]:
def search_time_based_feature_names(df):
    "각 시간별로 구성된 feature들(e.g. 'cloud_cover_0', 'cloud_cover_1' ...)에 대해 각각의 고유한 이름(e.g. 'cloud_cover')을 찾는다."

    time_based_features_pattern = re.compile(r"^(.*)_(\d{1,2})$")
    searched_time_based_feature_names = set()

    for column in df.columns:
        match = time_based_features_pattern.match(column)
        if match:
            feature_name, hour = match.groups()
            hour = int(hour)
            searched_time_based_feature_names.add(feature_name)

    return searched_time_based_feature_names

time_based_feature_names = search_time_based_feature_names(train_df)

print(len(time_based_feature_names))
print(time_based_feature_names)

14
{'surface_temp', 'sunshine_duration', 'sea_level_pressure', 'vapor_pressure', 'local_pressure', 'dew_point', 'wind_speed', 'min_cloud_height', 'wind_direction', 'visibility', 'cloud_cover', 'precipitation', 'snow_depth', 'humidity'}


train dataset에 있는 feature를 특성에 따라 다음과 같은 카테고리별로 정리하였다.

1. Data 식별
   - `id`: 순서(identical)
   - `station`: 지상관측소 번호(98: "동두천", 201: "강화" 등)
   - `station_name`: 지상관측소 이름("동두천", "서울", "강화" 등)
   - `date`: 날짜(월-일 형식, 1월 29일은 "01-29"로 표기)

2. 구름 관련 feature
    - `cloud_cover_[0-23]`: 증하층운량(10분위, 0~10)
    - `min_cloud_height_[0-23]`: 시간별 최저운고(100m 단위)

3. 온도 관련 feature
    - `dew_point_[0-23]`: 시간별 이슬점 온도(°C)
    - `surface_temp_[0-23]`: 시간별 지면온도(°C)
    - `climatology_temp`: 해당 날짜의 평균 기온(°C) (7년 동안 평균)

4. 습도 관련 feature
    - `humidity_[0-23]`: 시간별 습도(%)
    - `vapor_pressure_[0-23]`: 시간별 증기압(hPa)
    - `percipitation_[0-23]`: 시간별 강수량(mm)
    - `snow_depth_[0-23]`: 시간별 적설(cm)

5. 기압 관련 feature
    - `local_pressure_[0-23]`: 시간별 현지기압(hPa)
    - `sea_level_pressure_[0-23]`: 시간별 해면기압(hPa)

6. 가시성 관련 feature
    - `visibility_[0-23]`: 시간별 시정(10m 단위)
    - `sunshine_duration_[0-23]`: 시간별 일조(hr)

7. 바람 관련 feature
    - `wind_speed_[0-23]`: 시간별 풍속(m/s)
    - `wind_direction_[0-23]`: 시간별 풍향(도 단위)

8. target feature
    - `target`: 시간별 기온(°C, 다음날 평균 기온(°C)에서 `climatology_temp`를 뺀 값)


### 결측치 패턴

1. `-9999`: 관측소 기계에서 감지한 결측치 또는 이상치
2. `NaN`:
    - `sunshine_duration`: 야간 시간대(0, 1, 2, 3, 4, 5, 22, 23시), 즉 해가 없는 시간대
    - `snow_depth`: 눈이 오지 않은 경우
    - `precipitation`: 비가 오지 않은 경우




### Hourly Feature별 분석

#### Temperature-related Features

In [95]:
temp_prefixes = ['dew_point', 'surface_temp']
temp_cols = [col + "_" + str(hour) for col in temp_prefixes for hour in range(24)]
temp_cols.append('climatology_temp')